pkg/bootstrap/registries.go

```diff
@@ -0,0 +1,175 @@
+package bootstrap
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/pkg/errors"
+	"github.com/rancher/k3s/pkg/agent/templates"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+// Non-exported local type alias so that we can define new methods
+type registry templates.Registry
+
+// Explicit interface checks
+var _ authn.Keychain = &registry{}
+var _ http.RoundTripper = &registry{}
+
+// getPrivateRegistries loads private registry configuration from registries.yaml
+func getPrivateRegistries(path string) (*registry, error) {
+	privRegistries := &registry{}
+	privRegistryFile, err := ioutil.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	logrus.Infof("Using registry config file at %s", path)
+	if err := yaml.Unmarshal(privRegistryFile, &privRegistries); err != nil {
+		return nil, err
+	}
+	return privRegistries, nil
+}
+
+// Resolve returns an authenticator for the authn.Keychain interface. The authenticator
+// provides credentials to a registry by looking up configuration from mirror endpoints.
+func (r *registry) Resolve(target authn.Resource) (authn.Authenticator, error) {
+	endpointURL, err := r.getEndpointForHost(target.RegistryStr())
+	if err != nil {
+		return nil, err
+	}
+	return r.getAuthenticatorForEndpoint(endpointURL)
+}
+
+// RoundTrip round-trips a HTTP request for the http.RoundTripper interface. The round-tripper
+// overrides the Host in the headers and URL based on mirror endpoint configuration. It also
+// configures TLS based on the endpoint's TLS config, if any.
+func (r *registry) RoundTrip(req *http.Request) (*http.Response, error) {
+	endpointURL, err := r.getEndpointForHost(req.URL.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Host = endpointURL.Host
+	req.URL.Host = endpointURL.Host
+	req.URL.Scheme = endpointURL.Scheme
+
+	tlsConfig, err := r.getTLSConfigForEndpoint(endpointURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get TLS config from private registry configuration")
+	}
+
+	transport := &http.Transport{
```

Comment:
I think you need to CloseIdleConnections in defer, else disable the idle connections.

Reply:
This is just the default Transport from https://github.com/golang/go/blob/release-branch.go1.15/src/net/http/transport.go#L42. All I'm really doing beyond what you'd get from http.DefaultTransport is setting TLSClientConfig.

Since I'm not actually making the request, but rather letting the core code handle the actual round-tripping, I'm not responsible for closing anything, am I?

Reply:
Oh I see, I'm creating a new transport per request which means that idle connections can hang out on that Transport even though I'm never going to reuse it. I should either disable idle connections, or make sure that I return the same transport instance for the same endpoint.